### ISSUE TITLE
Enable custom server response headers through opensearch.yml configuration

### ISSUE THREAD

**Issue by asfoorial** [opensearch-project/OpenSearch#1069]
The objective of this feature is to provide a configuration open in opensearch.yml to enable the server to send custom HTTP response headers.

The option can be defined as shown below:
http.custom.headers: ["Strict-Security-Transport:max-age=31536000; includeSubDomains;"]

Running this "curl -v -u admin:admin http://localhost:9200" command would return the below bold header in the HTTP response. (The below response is coming from Elasticsearch but the same is expected with OpenSearch)
* Rebuilt URL to: http://localhost:9200/
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 9200 (#0)
* Server auth using Basic with user 'admin'
> GET / HTTP/1.1
> Host: localhost:9200
> Authorization: Basic YWRtaW46YWRtaW4=
> User-Agent: curl/7.55.1
> Accept: */*
>
< HTTP/1.1 200 OK
< _**Strict-Security-Transport: max-age=31536000; includeSubDomains;**_
< content-type: application/json; charset=UTF-8
< content-length: 506
<
{
  "name" : "LAPTOP-C0J5I3L2",
  "cluster_name" : "elasticsearch",
  "cluster_uuid" : "xVUOVOAGR4iaNHA2AEv1pw",
  "version" : {
    "number" : "7.8.0",
    "build_flavor" : "oss",
    "build_type" : "zip",
    "build_hash" : "unknown",
    "build_date" : "2021-02-16T10:30:42.633581400Z",
    "build_snapshot" : false,
    "lucene_version" : "8.5.1",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  },
  "tagline" : "You Know, for Search"
}

**Comment by dblock** [comment]
While we all love generic features that let users accomplish something quick and dirty, we should discuss whether it's the best course of action. There's a reason why web servers don't let you do that, e.g. [Tomcat](https://stackoverflow.com/questions/22972782/how-to-add-a-custom-header-in-tomcat) - it's too easy to shoot yourself in the foot by sending something incompatible or conflicting with a feature introduced in some future version. 

For generic use-cases I'd personally prefer a mechanism like a Tomcat filter, because it's code and isn't specific to adding headers (allows a lot more than headers). Alternatives could be recommending fronting OpenSearch with a proxy, for similar reasons.

For specific use-cases I'd prefer something less generic that doesn't accidentally conflict with some other feature and to think about how you might want to send this globally. In your example, setting `Strict-Security-Transport` might want to be a cluster-wide SSL option instead.

**Comment by dblock** [comment]
> I have created a local branch but I am not sure if I have permissions to push it. I highly appreciate help here.

That's expected. Use a GitHub fork and make all the changes on your fork, then PR them here. See https://guides.github.com/activities/forking/ for details.

**Comment by vrozov** [comment]
I don't think that HSTS is a good candidate for the **custom** server response header. The header should reflect the actual configuration and if TLS is enabled (currently can only be enabled by the security plugin), the HSTS header can be also added to the response. As a web browser is not the primary client for REST API and support for HSTS among other tools is quite limited and the security plugin does not support HTTP and HTTPS at the same time, I don't see a significant benefit of adding HSTS header to the server response.

**Comment by asfoorial** [comment]
HSTS is just one example. The feature might be needed for other cases as well. That is why it is available in OpenSearch Dashboards and Kibana.

I could not find a way to enable HSTS in the current version of OpenSearch. I can  use the custom header approach for now until it is implemented.

**Comment by vrozov** [comment]
> HSTS is just one example. The feature might be needed for other cases as well. That is why it is available in OpenSearch Dashboards and Kibana.

OpenSearch Dashboards and Kibana primary clients are web browsers, so is the support for HSTS. I am not aware of the ability to add custom headers to the Kibana.

> Besides, vulnerability scanning tools report the lack of HSTS as a security vulnerability that is why it is significant to have this header in the response.

There is no vulnerability, a scanning tool reports false positive.

**Comment by asfoorial** [comment]
@vrozov Kibana has this capability (server.customResponseHeaders) which is described in Kibana documentation here https://www.elastic.co/guide/en/kibana/7.x/settings.html

And my point about it being a security vulnerability, some companies require HSTS to be implemented for all their systems to enforce HTTPS across both servers and clients. Not sending HSTS header could create vulnerability that lead to man-in-the-middle attacks in such environments.

[HTTP Strict Transport Security (HSTS) is a policy mechanism that helps to protect websites against man-in-the-middle attacks such as protocol downgrade attacks and cookie hijacking](https://en.wikipedia.org/wiki/HTTP_Strict_Transport_Security)

**Comment by vrozov** [comment]
@asfoorial which client do you use to access OpenSearch REST API?

According to [rfc 6797](https://datatracker.ietf.org/doc/html/rfc6797#section-2.1) it covers:

   -  Web browser user wishes to interact with various web sites (some
      arbitrary, some known) in a secure fashion.

   -  Web site deployer wishes to offer their site in an explicitly
      secure fashion for their own, as well as their users', benefit.

OpenSearch is not a web site and is not an application server, it exposes **stateless** REST API over HTTP (without security plugin) or HTTPS (with security plugin).

**Comment by asfoorial** [comment]
If a customer requires HTTPS enforcement then surely OpenSearch will be used over HTTPS with security plugin configured. In this case, HSTS is applicable since there a communication over HTTPS. There is really no difference between a website, application server and REST API in this context.

And most importantly, I have a customer that requires it. I have implemented it and I am going to issue a PR soon.

Thanks to everyone.

**Comment by vrozov** [comment]
> If a customer requires HTTPS enforcement then surely OpenSearch will be used over HTTPS with security plugin configured. In this case, HSTS is applicable since there is communication over HTTPS. There is really no difference between a website, application server and REST API in this context.

HSTS is a hint/request to web browser not to use HTTP protocol for communicating with web site or application server when a response contains URL(s). The same does not apply to **stateless** REST API. Each request is fully independent so adding the header that is ignored/not supported by REST API clients has no impact on security.

**Comment by asfoorial** [comment]
This feature request is not about handling HSTS in specific but rather about offering the flexibility for OpenSearch admins to enable sending custom headers in the HTTP response when needed.

HSTS is one example that I needed it for. You may not be convinced that HSTS is applicable here but if it is required by a customer then we have not much of a choice but to satisfy it. 

Note: my customers mostly call OpenSearch REST APIs from the browser (Javascript).

eventhough the example is related to HSTS. I

**Comment by vrozov** [comment]
IMO, it will be better to explain customer that HSTS does not apply to OpenSearch as it is not a web application. Even if CX makes calls to REST API from a browser using Java script, URL used for calls is not a result of a call to OpenSearch and HSTS does not add any value. IMO, the use case for HSTS is not valid.

I'll leave it up to @dblock to see if adding support for generic custom headers is aligned with OpenSearch community design/vision.

**Comment by asfoorial** [comment]
Again Kibana & OpenSearch Dashboards already have that feature so it would be fair to have it also in OpenSearch.

**Comment by vrozov** [comment]
@dblock @nknize For HSTS, Authentication and any other security related headers, I'd prefer to manage them on the security plugin side until OpenSearch supports SSL and security to avoid any possible conflicts/misconfigurations.

**Comment by japrogramer** [comment]
I would like to control the content-type header with a custom header. Opensearch returns an array of values for this heather but the client, Hasura, does not support an array for this header. https://github.com/hasura/graphql-engine/issues/8598

**Comment by dblock** [comment]
@japrogramer Thanks. We're looking for a PR that can be merged (see https://github.com/opensearch-project/OpenSearch/pull/1096 for comments on a previous attempt).
